File: 351003/Razhnou/task361/publisher/src/main/java/by/bsuir/task361/publisher/service/CurrentUserService.java
package by.bsuir.task361.publisher.service;

import by.bsuir.task361.publisher.entity.User;
import by.bsuir.task361.publisher.entity.UserRole;
import by.bsuir.task361.publisher.exception.ApiException;
import org.springframework.http.HttpStatus;
import org.springframework.security.authentication.AnonymousAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.stereotype.Service;

@Service
public class CurrentUserService {
    private final UserService userService;

    public CurrentUserService(UserService userService) {
        this.userService = userService;
    }

    public String getCurrentLogin() {
        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
        if (authentication == null
                || !authentication.isAuthenticated()
                || authentication instanceof AnonymousAuthenticationToken) {
            throw new ApiException(HttpStatus.UNAUTHORIZED, 40101, "Authentication required");
        }
        return authentication.getName();
    }

    public UserRole getCurrentRole() {
        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
        if (authentication == null) {
            throw new ApiException(HttpStatus.UNAUTHORIZED, 40101, "Authentication required");
        }
        return authentication.getAuthorities().stream()
                .map(GrantedAuthority::getAuthority)
                .filter(authority -> authority.startsWith("ROLE_"))
                .findFirst()
                .map(authority -> UserRole.valueOf(authority.substring(5)))
                .orElseThrow(() -> new ApiException(HttpStatus.UNAUTHORIZED, 40101, "Authentication required"));
    }

    public User getCurrentUser() {
        return userService.getUserByLogin(getCurrentLogin());
    }

    public boolean isAdmin() {
        return getCurrentRole() == UserRole.ADMIN;
    }
}

File: 351002/Abrashin/publisher/src/main/kotlin/com/example/distcomp/security/CurrentUserService.kt
package com.example.distcomp.security

import com.example.distcomp.model.Creator
import com.example.distcomp.model.CreatorRole
import com.example.distcomp.repository.CreatorRepository
import org.springframework.security.access.AccessDeniedException
import org.springframework.security.authentication.AnonymousAuthenticationToken
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.stereotype.Service

@Service
class CurrentUserService(
    private val creatorRepository: CreatorRepository
) {
    fun currentLogin(): String {
        val authentication = SecurityContextHolder.getContext().authentication
        if (authentication == null || !authentication.isAuthenticated || authentication is AnonymousAuthenticationToken) {
            throw AccessDeniedException("Authentication required")
        }
        return authentication.name
    }

    fun currentRole(): CreatorRole {
        val authority = SecurityContextHolder.getContext().authentication?.authorities
            ?.firstOrNull()
            ?.authority
            ?.removePrefix("ROLE_")
            ?: throw AccessDeniedException("Role not found")
        return CreatorRole.valueOf(authority)
    }

    fun isAdmin(): Boolean = currentRole() == CreatorRole.ADMIN

    fun currentCreator(): Creator =
        creatorRepository.findByLogin(currentLogin())
            ?: throw AccessDeniedException("Authenticated creator not found")

    fun currentCreatorId(): Long =
        currentCreator().id ?: throw AccessDeniedException("Authenticated creator id not found")
}
